### ISSUE TITLE
It is very difficult to debug parsers

### ISSUE THREAD

**Issue by pbiggar** [elm-tools/parser#26]
tl;dr: can't debug parsers well, here's some things I tried.

I have a relatively straightforward parser that is currently about 170 lines, and it isn't working. It is already extremely difficult to figure out why my parser isn't working. A feature around this (or a guide explaining how to do this if you already have known good techniques) would be extremely useful.

Here are some things I've tried:

- The obvious place I have control over is the functions that are run after successful matches:

```
myInt = succeed (MyInt >> debug)
  |= ...
  |. ...
```

This showed me when something matched, which was slightly useful.

- Next, I added contexts to all my parsers, and put debugging into the Parser itself: annotating every call like `(parse state)`

```
printContext : String -> (State -> Step a) -> State -> Step a
printContext msg parse state = 
  let outputStrFn msg con = List.map (\c -> c.description ++ " (" ++ msg ++ "): " ++ (toString c.row) ++ ", " ++ (toString c.col)) con |> String.join "\n"
      outputFn c = let _ = Debug.log (outputStrFn msg c) () in c
      result = parse { state | context = (outputFn state.context)}
  in 
    case result of
      Good _ _ -> let _ = Debug.log ("GOOD (" ++ msg ++ ")") (state.context |> List.head |> Maybe.map .description) in result
      Bad _ _ -> let _ = Debug.log ("BAD (" ++ msg ++ ")") (state.context |> List.head |> Maybe.map .description) in result


{-| Run a parser *and then* run another parser!
-}
andThen : (a -> Parser b) -> Parser a -> Parser b
andThen callback (Parser parseA) =
  Parser <| \state1 ->
    case printContext "andThen2" parseA state1 of       -- <===== HERE
      Bad x state2 ->
        Bad x state2

      Good a state2 ->
        let
          (Parser parseB) =
            callback a
        in
          printContext "andThen2" parseB state2             -- <===== ALSO HERE
```

This does show me how my parsers behave and give me a rough structure of what's happening. Indeed this shows me that some of my parsers in my oneOf aren't being called, but I'm still not sure what's going on.

I think this tells me that I need to see what's happening in my pipelines too. My best hypothesis right now is that `oneOf` is broken, but I still don't have enough info to have any conviction in that.

- Next: annotating skips.

**Comment by pbiggar** [comment]
I eventually found my problem. I discovered that to really be able to debug it I needed to see:

- what matched and failed (so a lot of debug statements on `Good` and `Bad` case statements). I discovered a parser which shouldn't have failed, which was one of my bugs. I could possibly have worked this out by calling successively smaller parsers.

- the state transitions: I knew that my bug was something committing where I didn't want it to, but I couldn't figure out how to solve it. So I needed to annotate each place that returned a Step to figure out which state I wanted to alter (it was the last case in the `oneOfHelp`). I think I could have reasoned that if I knew enough about how the parser worked and how committing worked.


Overall, I think seeing the callgraph of my parser was essential to me figuring out what was going on. I have a hacked up version of the parser that I'm going to keep around because it is so useful. As such, I think that something that allows this (esp considering the lack of sourcemaps or a source debugger in Elm), is going to be really useful to be writing parsers using this library.

**Comment by evancz** [comment]
I do not get a clear sense of the feature request here, so I do not think it makes sense to keep tracking it. Next steps would be to refine this into a concrete idea, and ideally try an implementation to prove that it (1) is implementable, (2) actually helps debug, and (3) does not impose extreme performance penalties.

Having it be better to learn and debug is great, but there has to be a very specific plan to prove that it is viable and worth prioritizing.
